File: RotBoiRemastered/Systems/MetaProgression.cs
namespace RotBoiRemastered.Systems;

public sealed record SkillNode(
    string Key, string Symbol, string Name, string Description, string Stat,
    double ValuePerLevel, string Mode = "additive", int MaxLevel = 5, int BaseCost = 1);

public sealed record QuestDefinition(
    string Key, string Symbol, string Name, string Description, string Counter, long Target, int Reward = 1);

/// <summary>Permanent, UI-independent progression rules shared by the Soul and run startup.</summary>
public static class MetaProgression
{
    public const int StorageCapacity = 10;

    public static readonly IReadOnlyList<SkillNode> SkillNodes = new[]
    {
        new SkillNode("tempered_soul", "DMG", "Tempered Soul", "+2% damage per rank", "Bullet Damage", 1.02, "multiplicative"),
        new SkillNode("quick_memory", "SPD", "Quick Memory", "+2% movement speed per rank", "Player Speed", 1.02, "multiplicative"),
        new SkillNode("deep_reserve", "HP", "Deep Reserve", "+30 maximum health per rank", "Health", 30),
        new SkillNode("patient_hands", "CRT", "Patient Hands", "+1% critical chance per rank", "Crit Chance", .01),
        new SkillNode("wide_grasp", "AUR", "Wide Grasp", "+8 pickup range per rank", "Aura Size", 8),
        new SkillNode("soul_mail", "DEF", "Soul Mail", "+2 defense per rank", "Defense", 2),
        new SkillNode("steady_breath", "VIT", "Steady Breath", "+3 vitality per rank", "Vitality", 3),
        new SkillNode("far_memory", "RNG", "Far Memory", "+3% projectile range per rank", "Bullet Range", 1.03, "multiplicative"),
        new SkillNode("lively_hands", "RATE", "Lively Hands", "2% faster attacks per rank", "Attack Speed", .98, "multiplicative"),
        new SkillNode("keen_echo", "XP", "Keen Echo", "+3% experience per rank", "Exp Multiplier", 1.03, "multiplicative"),
        new SkillNode("heavy_echo", "SIZE", "Heavy Echo", "+2% projectile size per rank", "Bullet Size", 1.02, "multiplicative"),
        new SkillNode("straight_thought", "SHOT", "Straight Thought", "+3% projectile speed per rank", "Bullet Speed", 1.03, "multiplicative"),
    };

    public static readonly IReadOnlyDictionary<string, SkillNode> SkillNodesByKey = SkillNodes.ToDictionary(node => node.Key);

    public static readonly IReadOnlyList<QuestDefinition> Quests = new[]
    {
        new QuestDefinition("first_steps", "SKULL", "First Steps", "Defeat 50 enemies across any runs.", "enemies_defeated", 50),
        new QuestDefinition("crowd_control", "SKULL", "Crowd Control", "Defeat 250 enemies.", "enemies_defeated", 250),
        new QuestDefinition("harvester", "SKULL", "Harvester", "Defeat 1,000 enemies.", "enemies_defeated", 1000, 2),
        new QuestDefinition("curator", "CHEST", "Curator", "Discover 8 distinct items.", "items_found", 8),
        new QuestDefinition("full_shelf", "CHEST", "Full Shelf", "Discover 20 distinct items.", "items_found", 20, 2),
        new QuestDefinition("pathwalker", "PATH", "Pathwalker", "Complete 2 paths.", "path_clears", 2),
        new QuestDefinition("wayfinder", "PATH", "Wayfinder", "Complete 5 paths.", "path_clears", 5, 2),
        new QuestDefinition("afflictor", "DROP", "Afflictor", "Apply 100 status effects.", "statuses_applied", 100),
        new QuestDefinition("plague_hand", "DROP", "Plague Hand", "Apply 500 status effects.", "statuses_applied", 500, 2),
        new QuestDefinition("steady_fire", "SHOT", "Steady Fire", "Fire 1,000 projectiles.", "shots_fired", 1000),
        new QuestDefinition("storm", "SHOT", "Storm", "Fire 10,000 projectiles.", "shots_fired", 10000, 2),
        new QuestDefinition("keen_eye", "CRIT", "Keen Eye", "Land 100 critical hits.", "critical_hits", 100),
        new QuestDefinition("dead_center", "CRIT", "Dead Center", "Land 1,000 critical hits.", "critical_hits", 1000, 2),
        new QuestDefinition("hurt_the_rot", "DMG", "Hurt the Rot", "Deal 100,000 damage.", "damage_dealt", 100000),
        new QuestDefinition("break_the_rot", "DMG", "Break the Rot", "Deal 1,000,000 damage.", "damage_dealt", 1000000, 2),
        new QuestDefinition("boss_hunter", "CROWN", "Boss Hunter", "Defeat 3 bosses.", "bosses_defeated", 3),
        new QuestDefinition("boss_memory", "CROWN", "Boss Memory", "Defeat 15 bosses.", "bosses_defeated", 15, 2),
        new QuestDefinition("safe_return", "DOOR", "Safe Return", "Extract one run.", "runs_extracted", 1),
        new QuestDefinition("salvager", "DOOR", "Salvager", "Extract ten runs.", "runs_extracted", 10, 2),
        new QuestDefinition("student", "LEVEL", "Student", "Gain 25 levels.", "levels_gained", 25),
        new QuestDefinition("veteran", "LEVEL", "Veteran", "Gain 200 levels.", "levels_gained", 200, 2),
        new QuestDefinition("long_walk", "BOOT", "Long Walk", "Travel 10,000 world units.", "distance_traveled", 10000),
        new QuestDefinition("measured_force", "DPS", "Measured Force", "Deal 25,000 damage to the DPS effigy.", "dummy_damage", 25000),
        new QuestDefinition("living_weapon", "DPS", "Living Weapon", "Deal 250,000 damage to the DPS effigy.", "dummy_damage", 250000, 2),
    };

    public static void CompleteReadyQuests(bool save = true)
    {
        bool changed = false;
        foreach (var quest in Quests)
        {
            if (GameProfile.Profile.CompletedQuests.Contains(quest.Key)
                || GameProfile.Profile.QuestProgress.GetValueOrDefault(quest.Counter) < quest.Target)
                continue;
            GameProfile.Profile.CompletedQuests.Add(quest.Key);
            GameProfile.Profile.SoulTokens += quest.Reward;
            changed = true;
        }
        if (changed && save)
            GameProfile.SaveProfile();
    }

    public static bool PurchaseSkill(string key)
    {
        if (!SkillNodesByKey.TryGetValue(key, out var node))
            return false;
        int level = GameProfile.Profile.SkillLevels.GetValueOrDefault(key);
        int cost = node.BaseCost + level / 2;
        if (level >= node.MaxLevel || GameProfile.Profile.SoulTokens < cost)
            return false;
        GameProfile.Profile.SoulTokens -= cost;
        GameProfile.Profile.SkillLevels[key] = level + 1;
        GameProfile.SaveProfile();
        return true;
    }

    public static void ApplySkills(RunState state)
    {
        foreach (var (key, level) in GameProfile.Profile.SkillLevels)
        {
            if (!SkillNodesByKey.TryGetValue(key, out var node) || !state.Stats.TryGetValue(node.Stat, out var stat))
                continue;
            for (int rank = 0; rank < Math.Min(level, node.MaxLevel); rank++)
            {
                if (node.Mode == "multiplicative")
                    stat.Multiplicative.Add(node.ValuePerLevel);
                else
                    stat.Additive.Add(node.ValuePerLevel);
            }
        }
    }

    /// <summary>
    /// Writes the currently carried loadout (Equipment + Inventory) back to the
    /// profile so it survives past this run -- call whenever a run ends
    /// *without* dying (extract, complete, or a plain restart from the pause
    /// menu). GameSession.LoadCarriedItems is the inverse, reading this back
    /// into a fresh RunState.
    /// </summary>
    public static void SyncCarriedItems(RunState state)
    {
        GameProfile.Profile.CarriedEquipment = state.Equipment
            .Where(pair => pair.Value is not null)
            .ToDictionary(pair => pair.Key, pair => Items.Serialize(pair.Value!));
        GameProfile.Profile.CarriedInventory = state.Inventory
            .Select(item => item is null ? null : Items.Serialize(item))
            .ToList();
        GameProfile.SaveProfile();
    }

    /// <summary>The one and only "you died" path -- everything carried (not vaulted) is lost.</summary>
    public static void ClearCarriedItems()
    {
        GameProfile.Profile.CarriedEquipment.Clear();
        // Matches RunState.Inventory's fixed 8 slots (see its Reset()) -- kept as a literal
        // here too rather than a cross-layer reference to the UI's InventorySlotCount const.
        GameProfile.Profile.CarriedInventory = Enumerable.Repeat<StoredItemData?>(null, 8).ToList();
        GameProfile.SaveProfile();
    }

    public static void RecordExtraction(RunState state, string path, bool completed)
    {
        var run = new ExtractedRunData
        {
            Path = path,
            Outcome = completed ? "PATH COMPLETE" : "EXTRACTED",
            Level = state.CurrentLevel,
            Kills = state.NumOfEnemiesKilled,
            Seconds = state.RunTimeSeconds,
        };
        GameProfile.Profile.ExtractedRuns.Insert(0, run);
        if (GameProfile.Profile.ExtractedRuns.Count > 10)
            GameProfile.Profile.ExtractedRuns.RemoveRange(10, GameProfile.Profile.ExtractedRuns.Count - 10);
        GameProfile.IncrementQuest("runs_extracted");
        if (completed)
        {
            GameProfile.IncrementQuest("path_clears");
            GameProfile.Profile.PathMastery[path] = GameProfile.Profile.PathMastery.GetValueOrDefault(path) + 1;
        }
        GameProfile.SaveProfile();
    }
}

File: metaProgression.py
"""Small, UI-independent rules for Soul progression and persistent collections."""

from dataclasses import dataclass

import gameProfile
import items


@dataclass(frozen=True)
class SkillNode:
    key: str
    name: str
    description: str
    cost: int
    stat: str | None = None
    value: float = 0.0
    mode: str = "additive"


SKILL_NODES = (
    SkillNode("tempered_soul", "Tempered Soul", "+2% damage in every run", 1,
              "Bullet Damage", 1.02, "multiplicative"),
    SkillNode("quick_memory", "Quick Memory", "+2% movement speed", 1,
              "Player Speed", 1.02, "multiplicative"),
    SkillNode("deep_reserve", "Deep Reserve", "+30 maximum health", 1,
              "Health", 30),
    SkillNode("patient_hands", "Patient Hands", "+2% critical chance", 1,
              "Crit Chance", .02),
    SkillNode("wide_grasp", "Wide Grasp", "+8 experience attraction range", 1,
              "Aura Size", 8),
)
SKILL_NODES_BY_KEY = {node.key: node for node in SKILL_NODES}


QUESTS = (
    ("first_steps", "First Steps", "Defeat 50 enemies", "enemies_defeated", 50),
    ("curator", "Curator", "Find 8 distinct items", "items_found", 8),
    ("pathwalker", "Pathwalker", "Complete 2 paths", "path_clears", 2),
    ("afflictor", "Afflictor", "Apply 100 status effects", "statuses_applied", 100),
)


def stat_adjustments():
    additive, multiplicative = {}, {}
    for key in gameProfile.profile["skill_nodes"]:
        node = SKILL_NODES_BY_KEY.get(key)
        if node is None or node.stat is None:
            continue
        target = multiplicative if node.mode == "multiplicative" else additive
        target.setdefault(node.stat, []).append(node.value)
    return additive, multiplicative


def complete_ready_quests():
    changed = False
    completed = gameProfile.profile["completed_quests"]
    progress = gameProfile.profile["quest_progress"]
    for quest_id, _name, _description, counter, target in QUESTS:
        if quest_id not in completed and int(progress.get(counter, 0)) >= target:
            completed.append(quest_id)
            gameProfile.profile["soul_tokens"] += 1
            changed = True
    if changed:
        gameProfile.save_profile()
    return changed


def store_drop(drop):
    serialized = items.serialize(drop)
    storage = gameProfile.profile["storage"]
    # Every extracted copy matters.  Storage is an inventory, not a discovery
    # checklist; deduplicating identical rolls made one item effectively infinite.
    storage.append(serialized)
    gameProfile.discover_item(drop.name)


def extract_equipment(equipment):
    for drop in equipment.values():
        if drop is not None:
            store_drop(drop)
    gameProfile.save_profile()


def empty_equipment():
    return {"weapon": None, "armor": None, "ring": None,
            "accessory_1": None, "accessory_2": None}


def starting_equipment():
    """Preview the selected next-run kit without changing persistent storage."""
    result = empty_equipment()
    for slot, serialized in gameProfile.profile["starting_loadout"].items():
        if slot in result:
            result[slot] = items.deserialize(serialized)
    return result


def begin_run():
    """Withdraw the selected kit from storage and return its runtime equipment.

    Withdrawal happens before combat.  Extraction deposits surviving equipped
    items again; defeat, restart, or abandonment intentionally leaves them lost.
    Permanent skill nodes and other profile progression are never touched here.
    """
    equipment = empty_equipment()
    storage = gameProfile.profile["storage"]
    for slot, serialized in tuple(gameProfile.profile["starting_loadout"].items()):
        if slot not in equipment:
            continue
        try:
            storage_index = storage.index(serialized)
        except ValueError:
            continue
        drop = items.deserialize(storage.pop(storage_index))
        if drop is not None:
            equipment[slot] = drop
    gameProfile.profile["starting_loadout"] = {}
    gameProfile.save_profile()
    return equipment


def equip_from_storage(index):
    storage = gameProfile.profile["storage"]
    if not 0 <= index < len(storage):
        return False
    drop = items.deserialize(storage[index])
    if drop is None:
        return False
    loadout = gameProfile.profile["starting_loadout"]
    compatible = [drop.slot_type] if drop.slot_type != "accessory" else ["accessory_1", "accessory_2"]
    target = next((slot for slot in compatible if slot not in loadout), compatible[0])
    serialized = items.serialize(drop)
    # A selected loadout may not promise more identical copies than storage owns.
    already_selected = sum(1 for slot, value in loadout.items()
                           if slot != target and value == serialized)
    if already_selected >= storage.count(serialized):
        return False
    loadout[target] = serialized
    gameProfile.save_profile()
    return True


def clear_loadout():
    gameProfile.profile["starting_loadout"] = {}
    gameProfile.save_profile()


def mastery_title(path_key):
    count = int(gameProfile.profile["path_mastery"].get(path_key, 0))
    if count >= 5:
        return "Master"
    if count >= 3:
        return "Adept"
    if count >= 1:
        return "Initiate"
    return "Unwalked"
